background/lib/asset-similarity.ts

```diff
@@ -0,0 +1,115 @@
+import { AnyAsset, isSmartContractFungibleAsset } from "../assets"
+import { normalizeEVMAddress } from "./utils"
+
+/**
+ * Use heuristics to score two assets based on their metadata similarity. The
+ * higher the score, the more likely the asset metadata refers to the same
+ * asset.
+ *
+ * @param a - the first asset
+ * @param b - the second asset
+ * @return an integer score >= 0
+ */
+export function scoreAssetSimilarity(a: AnyAsset, b: AnyAsset): number {
+  let score = 0
+  if (a.symbol === b.symbol) {
+    score += 1
+  }
+  if (a.name === b.name) {
+    score += 1
+  }
+  if ("decimals" in a && "decimals" in b) {
+    if (a.decimals === b.decimals) {
+      score += 1
+    } else {
+      score -= 1
+    }
+  } else if ("decimals" in a || "decimals" in b) {
+    score -= 1
+  }
```

Comment:
```suggestion
  if ("decimals" in a || "decimals" in b) {
    if (a.decimals === b.decimals) {
      score += 1
    } else {
      score -= 1
    }
  }
```

Reply:
Unfortunately, this won't pass type checking, as `a.decimals` is illegal if `"decimals" in a` is false. We could do this, however:

```suggestion
  if ("decimals" in a && "decimals" in b && a.decimals === b.decimals) {
      score += 1
  } else if ("decimals" in a || "decimals" in b) {
    score -= 1
  }
```

(Modulo prettier reformatting.) I do think that would be clearer overall.